### PR TITLE
Add clang tidy review in Lint CI

### PR DESCRIPTION
**Pull request by jcwchen**
### Description

Add clang tidy review developed by ZedThree/clang-tidy-review in Lint CI.

### Motivation and Context

Closes https://github.com/onnx/onnx/issues/4459.

### CODE REVIEW

File: onnx/shape_inference/implementation.cc [hunk 1]
```diff
@@ -1007,7 +1007,7 @@ std::vector<const TypeProto*> GraphInferencerImpl::doInferencing(
   return graph_output_types;
 }
 
-std::string GetErrorWithNodeInfo(const NodeProto& n, std::runtime_error err) {
+std::string GetErrorWithNodeInfo(const NodeProto& n, std::runtime_error err ) {
```
Comment (jcwchen) [inline onnx/shape_inference/implementation.cc:1010]: FYI: Just for experiment. Will remove it before PR is merged.

File: .github/workflows/lint.yaml [hunk 2]
```diff
@@ -36,6 +36,19 @@ jobs:
...
           filter_mode: diff_context
+      - name: clang-tidy review
+        uses: ZedThree/clang-tidy-review@ddfe311e5d2ee5ea096cedc21adcbdb83c32f62b # v0.12.1
+        with:
+          build_dir: '.setuptools-cmake-build'
+          apt_packages: libprotobuf-dev protobuf-compiler
+          lgtm_comment_body: ''
+          exclude: "/third_party/*"
+          split_workflow: true
+        id: review
```
[the remark this answers is not in the document]
Comment (jcwchen) [reply .github/workflows/lint.yaml:48]: It's a valid concern. In that case I won't let clang-tidy check fail any CI, but instead have clang-tidy review every PRs and comment. The author can decide whether to take advice or not on their own.

**Comment by justinchuby** [comment]
image

looking good

**Review by jcwchen — COMMENTED** [review]
Please note that we can only test this new CI after the CI file is included in the main branch. At least, I have tested this CI in my forked repo and it seems working.

**Comment by justinchuby** [comment]
Was there any permission we need to explicitly enable in the yaml? Since you have write access I thought it should work for you

**Comment by jcwchen** [comment]
> Was there any permission we need to explicitly enable in the yaml? Since you have write access I thought it should work for you

Good question. See the document from the author: https://github.com/marketplace/actions/clang-tidy-review#usage-in-fork-environments-split-workflow. It seems the PR coming from the forked one cannot post a comment before merging into the main branch. My experiment also proves that.

**Review by liqunfu — APPROVED** [review]

**Pull request merged**